core/src/data_sources/data_source.rs

```diff
@@ -730,25 +793,135 @@ impl DataSource {
             })
             .buffer_unordered(16)
             .map(|r| match r {
-                Err(e) => Err(anyhow!("DataSource document retrieval error: {}", e))?,
+                Err(e) => Err(anyhow!("datasource document retrieval error: {}", e))?,
                 Ok(r) => r,
             })
             .try_collect::<Vec<_>>()
             .await?;
 
-        // Merge the chunks with the documents.
-        let mut documents = documents
-            .into_iter()
+        let l_qdrant_client = Arc::new(Mutex::new(qdrant_client));
+        let mut documents = futures::stream::iter(documents)
             .map(|mut d| {
-                let chunks = chunks
+                let mut chunks = chunks
                     .iter()
                     .filter(|(document_id, _)| document_id == &d.document_id)
                     .map(|(_, c)| c.clone())
                     .collect::<Vec<Chunk>>();
-                d.chunks = chunks;
-                d
+                // look around a bit for arc/mutex stuff but I don't know if this is the right way to do this
+                let qdrant_client = l_qdrant_client.clone();
+                let collection = self.qdrant_collection();
+                let chunk_size = self.config.max_chunk_size;
+                tokio::spawn(async move {
+                    if target_document_tokens != 0 {
+                        let qdrant_client = qdrant_client.lock().await;
+                        let mut offset_set = std::collections::HashSet::new();
+                        for chunk in chunks.iter() {
+                            offset_set.insert(chunk.offset);
+                        }
+                        let current_length = chunks.iter().map(|c| c.text.len()).sum::<usize>();
```

Comment:
This is text length or you're comparing it below to a tokens length.

You want chunks*max_chunk_size (which might not be perfectly accurate at the boundaries but we have the guarantee that the number of tokens in chunks is less than max_chunk_size) so we won't ever go over `target_documents_tokens`.